Block params

JMMackenzie opened the pull request:
Some potential issues:
- I am not particularly familiar with the `boost::variant` stuff, so please check that I am not abusing it.
- I was able to handle most of the command line argument interactions using the CLI `excludes`, but this does not work for supplying `--lambda` without `--variable-block` so I wrote an explicit case to catch that one.

Also updated the documentation. Please suggest changes and I will try to fix ASAP.

Commit:
First attempt at allowing fixed block size args

Commit:
Docs and flag change

JMMackenzie commented:
Not currently building due to the case where `--variable-block` is passed without `--lambda` - I was passing the variant in as a default `uint64_t` so when it gets to the variable block partition and tries to `get()` the float, it fails (since the `uint64_t` is the set type). 

Not sure what the clean way to fix this is, do you have suggestions?

Commit:
Fix default param issue

JMMackenzie commented:
OK I have fixed the `test` issue, but I think this is quite ugly now (and I had to change the test cases to explicitly use the correct `variant` to avoid issues. Let me know what you think, we might want to refactor this.

elshize commented on include/pisa/wand_data.hpp:
```diff
@@ -1,5 +1,6 @@
 #pragma once
```
Let's `clang-format` these modified files, they're quite a mess.

elshize commented on include/pisa/wand_data.hpp:
```diff
@@ -22,7 +23,7 @@ namespace pisa {
         template<typename LengthsIterator>
         wand_data(LengthsIterator len_it, uint64_t num_docs,
-                      binary_freq_collection const &coll, partition_type type = partition_type::fixed_blocks, const float lambda = 0.0f) {
+                      binary_freq_collection const &coll, partition_type type = partition_type::fixed_blocks, boost::variant<float, uint64_t> block_size = uint64_t(0)) {
```
In general, I'd like to use `std::variant` instead. But seeing as there are some problems with its support by AppleClang (lol, I know), we might wanna use boost for now. @amallia Are we ready to use `std::variant` properly or not yet?

amallia replied:
Not yet, I asked @JMMackenzie to use `boost` because of the incompatibility.
We can migrate all the `boost::variant` occurrences all together at some point.

elshize commented on include/pisa/wand_data.hpp:
```diff
@@ -22,7 +23,7 @@ namespace pisa {
         template<typename LengthsIterator>
         wand_data(LengthsIterator len_it, uint64_t num_docs,
-                      binary_freq_collection const &coll, partition_type type = partition_type::fixed_blocks, const float lambda = 0.0f) {
+                      binary_freq_collection const &coll, partition_type type = partition_type::fixed_blocks, boost::variant<float, uint64_t> block_size = uint64_t(0)) {
```
Correct my if I'm wrong, but I think now that we have a variant type, `partition_type` is redundant, right?

amallia replied:
This is correct! and it is such a great improvement :)

JMMackenzie replied:
Refactored in the latest.

elshize commented on include/pisa/wand_data_compressed.hpp:
```diff
@@ -111,12 +112,12 @@ namespace {
 
-                    auto t = ((type == partition_type::fixed_blocks) ? static_block_partition(seq, norm_lens)
-                                                      : variable_block_partition(coll, seq, norm_lens, lambda));
+                    auto t = ((type == partition_type::fixed_blocks) ? static_block_partition(seq, norm_lens, boost::get<uint64_t>(block_size))
```
So typically, you would use a static visitor: https://www.boost.org/doc/libs/1_70_0/doc/html/variant/reference.html#variant.concepts.static-visitor

However, it is way too verbose for this, at least for inline use. Might as well do an `if` statement. If `type` not available, you can use `which` method that returns the index of the type that's in the variant. Not a great design though.

Not that for `std::variant`, it is very easy to write a wrapper function that turns this use case into:
```cpp
auto t = match(block_size,
    [](float alpha) { ... },
    [](uint64_t size) { ... })
```

You can see the example of that in on of the old commits in `warcpp`: https://github.com/pisa-engine/warcpp/blob/e8c53c09cb983ff6166360d46c514d8e36acb91a/include/warcpp/warcpp.hpp#L94 This **doesn't work** for our currently used AppleClang, though (blame Antonio).

It might be also possible for `boost::variant` but not sure how easy it is (the interface is slightly different). There is a lib for that: https://github.com/exclipy/inline_variant_visitor but it seems much longer compared to the `std::variant` version. And clearly, bringing another library for this is an overkill.

amallia replied:
I agree with this one!

JMMackenzie replied:
Yes, good point. I have used a similar idea that you mentioned below since we are using `boost` for now. Let's revisit this in #188 when we get to it.

elshize commented on include/pisa/wand_utils.hpp:
```diff
@@ -11,10 +11,13 @@ enum class partition_type { fixed_blocks, variable_blocks };
   std::vector<uint32_t> block_docid;
   std::vector<float> block_max_term_weight;
   uint64_t block_size = configuration::get().block_size;
+  if (in_block_size > 0)
```
Missing braces.

elshize commented on include/pisa/wand_utils.hpp:
```diff
@@ -11,10 +11,13 @@ enum class partition_type { fixed_blocks, variable_blocks };
 static_block_partition(binary_freq_collection::sequence const &seq,
-                       std::vector<float> const &norm_lens) {
+                       std::vector<float> const &norm_lens,
+                       const uint64_t in_block_size = 0) {
```
I think it would be more clear if we used `std::optional` here, explicitly indicating when a block size is defined and when it's not, instead of using a magic value 0:

```cpp
std::optional<uint64_t> const = std::nullopt) {
```

JMMackenzie replied:
I have decided it's better to not allow default parameters here. Instead, I have built the default constructor for either the `VariableBlock` or `FixedBlock` type to directly read the corresponding value from the config file. So, instantiating a default object will give us the default configuration, which can be used directly. Please check it out and let me know what you think - I believe this approach is much more robust and readable.

elshize commented on src/create_wand_data.cpp:
```diff
@@ -9,30 +10,32 @@
-    app.add_option("-l,--lambda", lambda, "Lambda parameter for variable blocks");
     auto var_block_opt = app.add_flag("--variable-block", variable_block, "Variable length blocks");
+    auto var_block_param_opt = app.add_option("-b,--block-size", fixed_block_size, "Block size for fixed-length blocks")->excludes(var_block_opt);
+    app.add_option("-l,--lambda", lambda, "Lambda parameter for variable blocks")->excludes(var_block_param_opt);
```
Add `->needs(var_block_opt)`

elshize commented on src/create_wand_data.cpp:
```diff
@@ -42,20 +45,36 @@ int main(int argc, const char **argv) {
     binary_freq_collection coll(input_basename.c_str());
 
+    // Initialize the variant to the correct type
+    boost::variant<float, uint64_t> block_size = uint64_t(0);
```
We probably want to make it `const`, so I'd either use another function to set this up, or just wrap it in a lambda:

```cpp
auto const block_size = [&] () -> boost::variant<float, uint64_t> {
    if (variable_block) {
        return lambda;
    } else {
        return fixed_block_size;
    }
}();
```
Not that I moved checking compliance of defined options above to the CLI part.

elshize reviewed (changes requested):
If you ask me, we should probably go with `std::variant`. If we're not ready to use `match`, at least we can use `std::get_if` to easily check for the type right now.

I would also create a `BlockSize` type defined as follows:
```cpp
struct FixedSize {
    uint64_t size;
};
struct VariableSize {
    float alpha;
};
using BlockSize = std::variant<FixedSize, VariableSize>;
```

Then, we could use it one of these ways:

1. If we upgrade AppleClang
```cpp
auto t = match(block_size,
    [](FixedSize size){ return f(size); },
    [](VariableSize size){ return g(size); });
```
2. If we don't
```cpp
auto t = std::get_if<FixedSize>(block_size) != nullptr
    ? f(*std::get_if<FixedSize>(block_size))
    : g(*std::get_if<VariableSize>(block_size))
```

amallia commented:
Thanks @JMMackenzie for the PR! I agree with all the comments @elshize already made.
Once those are addressed we can merge!

elshize commented:
So if we stick to `boost::variant`, I think something like this would work:
```cpp
auto t = block_size.type() == typeid(FixedSize)
    ? f(boost::get<FixedSize>(block_size))
    : g(boost::get<VariableSize>(block_size))
```
Needs to be double-checked, not sure if the standard `typeid` works here or must use some boost version, but I imagine this should work.

Also, I suggest to use the proposed "inner" types (see https://github.com/pisa-engine/pisa/pull/187#pullrequestreview-233460491) in function signatures, e.g.:

```cpp
static_block_partition(binary_freq_collection::sequence const &seq,
                       std::vector<float> const &norm_lens) {
                       std::vector<float> const &norm_lens,
                       std::optional<FixedSize> const in_block_size = std::nullopt) {
```

Commit:
Take two - refactor

JMMackenzie commented:
I have overhauled this now:

- Refactored the code to remove the `partition_type`
- Provide structs for each type that ingest default params out of config file
- Clean up using `clang_format` - I am not sure it did what it was expected to do, could you double check this?
- Added the CLI `needs` to `lambda` which improves simplicity of the `.cpp` file.

I think that's it for now. The code is greatly simplified, and seems to be working on my end.

elshize commented on include/pisa/wand_data_compressed.hpp:
```diff
@@ -9,230 +10,205 @@
 
-            float add_sequence(binary_freq_collection::sequence const &seq, binary_freq_collection const &coll, std::vector<float> const & norm_lens, const float lambda = 0.0f){
+                auto t =
+                    (block_size.type() == typeid(FixedBlock))
```
Nitpick: no need for the outer parens.

elshize commented on src/create_wand_data.cpp:
```diff
@@ -9,53 +10,68 @@
+{
     using namespace pisa;
 
     float lambda = 0.0f;
```
Let's use `std::optional<float> lambda{};` here instead. It works out of the box. The difference is that now it will be `std::nullopt` when no option is passed. Then it makes more sense when you check for that later.

JMMackenzie replied:
Yes, great point!

elshize commented on src/create_wand_data.cpp:
```diff
@@ -9,53 +10,68 @@
+{
     using namespace pisa;
 
     float lambda = 0.0f;
```
Let's use `std::optional<float> lambda{};` here instead. It works out of the box. The difference is that now it will be `std::nullopt` when no option is passed. Then it makes more sense when you check for that later.

elshize commented on src/create_wand_data.cpp:
```diff
@@ -9,53 +10,68 @@
     using namespace pisa;
 
     float lambda = 0.0f;
+    uint64_t fixed_block_size = 0;
```
Same as above.

JMMackenzie replied:
Got it, ta.

elshize commented on src/create_wand_data.cpp:
```diff
@@ -9,53 +10,68 @@
+    binary_collection sizes_coll((input_basename + ".sizes").c_str());
     binary_freq_collection coll(input_basename.c_str());
 
+    // Initialize the variant to the correct type, default arguments from config
```
How about we use the following code here. For one, it makes the block size `const`, and then also is clear. Rule of thumb: if you have to describe a specific block/branch in a comment, you most likely want to rearrange things to make it more clear---and then remove the comments.

```cpp
auto const block_size = [&]() -> BlockSize {
    if (variable_block) {
        return fixed_block_size ? FixedBlock(*fixed_block_size) : FixedBlock();
    } else {
        return lambda ? VariableBlock(*lambda) : VariableBlock();
    }
}();
```

JMMackenzie replied:
This is elegant, thanks! Incorporating it now.

elshize reviewed (changes requested):
@JMMackenzie make sure all tests are passing, seems like you haven't updated all the call sites.

JMMackenzie commented:
Good pickups, thanks a lot. I am pushing a new commit now which should rectify all of these issues (all tests pass on my system).

Commit:
Fix tests, improve readability

amallia committed:
Missing include
